tools/kagami/src/crypto.rs

```diff
@@ -1,14 +1,169 @@
+            Self::get_key_raw(&self.private_key_path, &self.private_key_string)?;
+            match private_key_bytes {
+                KeyStorage::FromFile(bytes) => PrivateKey::from_bytes(self.algorithm.0, bytes.as_slice()).wrap_err_with(|| {
+                    eyre!(
+                        "Failed to parse private key from bytes for algorithm `{}`",
+                        self.algorithm
+                    )
+                }),
+                KeyStorage::FromCLI(hex) => PrivateKey::from_hex(self.algorithm.0, hex).wrap_err_with(|| {
+                    eyre!(
+                        "Failed to parse private key from hex for algorithm `{}`",
+                        self.algorithm
+                    )
+                }),
+            }
+    }
+
+    fn get_public_key(&self) -> Result<PublicKey> {
+        let public_key_bytes = Self::get_key_raw(&self.public_key_path, &self.public_key_string)?;
+        match public_key_bytes {
+            KeyStorage::FromFile(bytes) => PublicKey::from_bytes(self.algorithm.0, bytes.as_slice()).wrap_err_with(|| {
+                eyre!(
+                    "Failed to parse public key from bytes for algorithm `{}`",
+                    self.algorithm
+                )
+            }),
+            KeyStorage::FromCLI(hex) => {
+                if hex.starts_with(ED25519_PREFIX) {
+                    PublicKey::from_str(hex).wrap_err_with(|| {
+                        eyre!(
+                            "Failed to deserialize public key from hex for algorithm `{}`",
+                            self.algorithm
+                        )
+                    })
+                } else {
+                    PublicKey::from_hex(self.algorithm.0, hex).wrap_err_with(|| {
+                    eyre!(
+                        "Failed to parse public key from hex for algorithm `{}`",
+                        self.algorithm
+                    )
+                })
+            }
+        },
+        }
+    }
+
+    fn get_key_raw<'a, P: AsRef<Path>>(
+        path: &Option<P>,
+        value: &'a Option<String>,
+    ) -> Result<KeyStorage<'a>, std::io::Error> {
+        match (path, value) {
+            (Some(path_buf), None) => Ok(KeyStorage::FromFile(fs::read(path_buf)?)),
+            (None, Some(hex)) => Ok(KeyStorage::FromCLI(hex.as_str())),
+            _ => unreachable!("Clap group invariant"),
+        }
+    }
+
+    fn read_keypair<P: AsRef<Path>>(path: P) -> Result<KeyPair> {
+        let bytes = fs::read(path)?;
+        Ok(serde_json::from_slice(bytes.as_slice())?)
+    }
+}
+
+impl<T: Write> RunArgs<T> for GenesisSigningArgs {
+    fn run(self, writer: &mut BufWriter<T>) -> Outcome {
+        let key_pair = if let Some(path) = self.keypair_path {
+            Self::read_keypair(path)?
+        } else {
+            let (public_key, private_key) = (self.get_public_key()?, self.get_private_key()?);
+            KeyPair::new(public_key, private_key)?
+        };
```

Comment:
If we always require `KeyPair` maybe we can simplify code to only allow keypair?

Reply:
I thought that it might be complex to pass KeyPair as a command line argument. And I thought that we don't want to enforce user to pass it through the file, so I decided to support both options

Reply:
I agree. Having both a `--key-pair` and `--private-key`/`--public-key` is good for CLI, IMO.

Although, I just realized that `--public-key` is redundant, as it might be derived from the private key.

Reply:
> Although, I just realized that --public-key is redundant, as it might be derived from the private key. Maybe even having only --private-key is enough.

it cannot always be derived depending on which algorithm is used

Reply:
> I agree. Having both a --key-pair and --private-key/--public-key is good for CLI, IMO.

yes if they are mutually exclusive

Reply:
They are mutually exclusive which is enforced by clap grouping. I think it's pretty convenient to have keypair in json and load it from the file. And the same goes for passing public and private key to cli

Reply:
why is public key ever necessary when signing genesis?

Reply:
I see it's because of the fact that we pack public keys with signatures